### PR TITLE
99 signup

### PR DESCRIPTION
**Pull request by devstm**
### Description:

**_create an Express Route to create new user in web site._**

1. validate user info request
2. create new user in database
3. send user id and name with encrypted Cookies
4. test the route

### Required middle wares:
1. custumError

Request body:
```
{
  first_name,
  last_name,
  email,
  password,
  phone,
  whatsapp,
  locationId,
  image,
  description,
  facebook_link,
  instagram_link,
  mainServiceId,
}
```

Validation
```
{
  first_name: Joi.string().required(),
  last_name: Joi.string().required(),
  email: Joi.string().email().required(),
  password: Joi.string().pattern(/^[a-zA-Z0-9]{8,30}$/),
  phone: Joi.string().required(),
  whatsapp: Joi.string().required(),
  locationId: Joi.number().required(),
  image: Joi.string().required(),
  description: Joi.string(),
  facebook_link: Joi.string(),
  instagram_link: Joi.string(),
  mainServiceId: Joi.number().required(),
}
```

Tests:
 201 => Post /api/v1/signup return with status code equal 201 created
 400 => Post /api/v1/signup return with status code equal 400 duplicated Email
 400 => Post /api/v1/signup return with status code equal 400 duplicated Phone
 400 => Post /api/v1/signup return with status code equal 400 invalid email
 400 => Post /api/v1/signup return with status code equal 400 without first name 
 400 => Post /api/v1/signup return with status code equal 400 with invalid password

### CODE REVIEW

**Review by RandSohail — APPROVED** [review]

File: server/utils/validation/signUpValidation.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+const signUpValidation = Joi.object({
+  first_name: Joi.string().required(),
+  last_name: Joi.string().required(),
+  email: Joi.string().email().required(),
+  password: Joi.string().pattern(/^[a-zA-Z0-9]{8,30}$/),
+  phone: Joi.string().required(),
+  whatsapp: Joi.string().required(),
+  locationId: Joi.number().required(),
+  image: Joi.string().required(),
```
Comment (LinaYahya) [inline server/utils/validation/signUpValidation.js:11]: I guess we will not send an image in the signup route

File: server/utils/validation/signUpValidation.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  description: Joi.string(),
+  facebook_link: Joi.string(),
+  instagram_link: Joi.string(),
+  mainServiceId: Joi.number().required(),
```
Comment (LinaYahya) [inline server/utils/validation/signUpValidation.js:15]: and sub-services which is not required

File: server/controllers/signup/index.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    const userData = await signUpValidation.validateAsync(req.body);
+    const message = await checkIsUserExists(userData);
+    if (message) throw customError(message, 400);
+    const dataToInsert = await getDataToInsert(userData);
+    const user = await User.create(dataToInsert);
+    const { token, data } = await generateToken(user);
+    res.status(201).cookie('userToken', token).json({ msg: 'logged in successfully', data });
+  } catch (err) {
+    if (err.name === 'ValidationError') {
+      next(customError(err.message, 400));
```
Comment (LinaYahya) [inline server/controllers/signup/index.js:19]: either to return or to add else block for next error

File: server/controllers/signup/dataToInsert.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+const { hash } = require('bcrypt');
+
+const getDataToInsert = async (data) => {
+  const { password, whatsapp } = data;
+  const whatsappNumber = whatsapp.slice(2);
```
Comment (LinaYahya) [inline server/controllers/signup/dataToInsert.js:5]: leave a comment about the shape of WhatsApp number

File: server/controllers/signup/index.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+const { generateToken } = require('./generateToken');
+const { getDataToInsert } = require('./dataToInsert');
+
+const signUp = async (req, res, next) => {
+  try {
+    const userData = await signUpValidation.validateAsync(req.body);
+    const message = await checkIsUserExists(userData);
+    if (message) throw customError(message, 400);
+    const dataToInsert = await getDataToInsert(userData);
+    const user = await User.create(dataToInsert);
```
Comment (LinaYahya) [inline server/controllers/signup/index.js:14]: find a way to insert subservices

**Review by LinaYahya — CHANGES_REQUESTED** [review]
Great Job والله 🥳

**Review by LinaYahya — APPROVED** [review]

**Review by abdulrahman-2020 — APPROVED** [review]

**Pull request merged**